File: wood_factory/wood_factory/page/factory_scan/factory_scan.js
frappe.pages["factory-scan"].on_page_load = function (wrapper) {
    const page = frappe.ui.make_app_page({parent: wrapper, title: __("Factory Scan"), single_column: true});
    page.main.addClass("factory-scan-page");
    page.main.html(`<div class="scan-panel"><h2>${__("Scan Factory Code")}</h2><p class="text-muted">${__("Scan the whole order code. Only exception pieces are opened separately.")}</p><input class="form-control scan-input" type="text" autocomplete="off" placeholder="WF:FO:... / WF:FP:..."><button class="btn btn-primary scan-submit">${__("Open Work")}</button></div><div class="scan-result"></div>`);
    const input = page.main.find(".scan-input");
    page.main.find(".scan-submit").on("click", () => resolve_scan(page));
    input.on("keydown", event => { if (event.key === "Enter") resolve_scan(page); });
    setTimeout(() => input.trigger("focus"), 100);
};

function resolve_scan(page) {
    const input = page.main.find(".scan-input"), code = input.val().trim();
    if (!code) return;
    frappe.call({method: "wood_factory.wood_factory.page.factory_scan.factory_scan.resolve_scan", args: {code}, freeze: true}).then(r => {
        const result = r.message || {};
        render_scan_result(page, result);
        input.val("").trigger("focus");
    });
}

function render_scan_result(page, result) {
    const box = page.main.find(".scan-result");
    const title = result.kind === "order" ? result.name : (result.part_name || result.piece_uid);
    const detail = result.kind === "order" ? `${result.customer || ""} · ${result.progress_percent || 0}%` : `${result.factory_order} · ${result.width_mm} × ${result.height_mm} mm`;
    box.html(`<div class="scan-card ${result.kind === "piece" ? "scan-exception" : ""}"><div><h3>${frappe.utils.escape_html(title || "")}</h3><div>${frappe.utils.escape_html(detail)}</div><strong>${__(result.stage || "")} · ${__(result.status || "")}</strong>${result.message ? `<p>${__(result.message)}</p>` : ""}</div><button class="btn btn-primary scan-open">${__("Open Work")}</button></div>`);
    box.find(".scan-open").on("click", () => frappe.set_route(...result.route));
}

File: wood_factory/wood_factory/page/factory_scan/factory_scan.py
import frappe

from wood_factory.security import require_operational_view


@frappe.whitelist()
def resolve_scan(code):
    require_operational_view()
    code = (code or "").strip()
    if not code:
        frappe.throw("Scan or enter a factory code")

    order_name = _extract_code(code, "FO")
    piece_uid = _extract_code(code, "FP")

    order = _find_order(order_name or code)
    if order:
        _require_read("Factory Order", order)
        return _order_result(order)

    piece = _find_piece(piece_uid or code)
    if piece:
        _require_read("Factory Piece", piece)
        if not piece.is_exception:
            order = frappe.get_doc("Factory Order", piece.factory_order)
            _require_read("Factory Order", order)
            result = _order_result(order)
            result["message"] = "This is a normal piece. Work is controlled by the whole Factory Order."
            return result
        return _piece_result(piece)

    frappe.throw(f"No Factory Order or Factory Piece matches code {code}")


def _require_read(doctype, doc):
    if not frappe.has_permission(doctype, ptype="read", doc=doc):
        frappe.throw(f"You are not permitted to view this {doctype}", frappe.PermissionError)


def _extract_code(value, prefix):
    marker = f"WF:{prefix}:"
    return value.split(marker, 1)[1].strip() if marker in value else None


def _find_order(value):
    name = frappe.db.exists("Factory Order", value)
    return frappe.get_doc("Factory Order", name) if name else None


def _find_piece(value):
    name = frappe.db.exists("Factory Piece", value) or frappe.db.get_value("Factory Piece", {"piece_uid": value}, "name")
    return frappe.get_doc("Factory Piece", name) if name else None


def _order_result(order):
    stage = next((row for row in order.production_stages if row.stage == order.current_stage), None)
    return {
        "kind": "order",
        "name": order.name,
        "scan_code": f"WF:FO:{order.name}",
        "customer": order.customer,
        "stage": order.current_stage,
        "status": stage.status if stage else order.status,
        "progress_percent": order.progress_percent,
        "route": ["factory-worker"],
    }


def _piece_result(piece):
    return {
        "kind": "piece",
        "name": piece.name,
        "scan_code": f"WF:FP:{piece.piece_uid}",
        "piece_uid": piece.piece_uid,
        "factory_order": piece.factory_order,
        "part_name": piece.part_name,
        "width_mm": piece.width_mm,
        "height_mm": piece.height_mm,
        "stage": piece.current_stage,
        "status": piece.status,
        "route": ["Form", "Factory Piece", piece.name],
    }
